### DISCUSSION TITLE
How to update docker installation?

### DISCUSSION THREAD

**Discussion by arladmin** [Budibase/budibase#1198] [Help] [answered]
Hi all.

My setup: DigitalOcean Docker droplet. Budibase set up using this [procedure](https://docs.budibase.com/self-hosting/hosting-methods/docker-compose#setting-up-budibase)

**How to update a running setup?**
I'm facing some issues with this.

The way i usually update my other docker based projects is:
1. Pull the latest images for the appropriate service(s)
2. `docker-compose stop`
3. `docker-compose rm`
4. `docker-compose up -d`

When i tried `docker-compose stop` in my budibase setup, i see these errors:
`ERROR: The Compose file './docker-compose.yaml' is invalid because:`
`services.proxy-service.ports contains an invalid type, it should be a number, or an object`
`services.app-service.ports contains an invalid type, it should be a number, or an object`
`services.worker-service.ports contains an invalid type, it should be a number, or an object`
`services.couchdb-service.ports contains an invalid type, it should be a number, or an object`
`services.minio-service.ports contains an invalid type, it should be a number, or an object`

I was finally able to stop and delete the existing containers with these commands:
`docker stop $(docker ps -a -q) `
`docker rm $(docker ps -a -q)`

and to bring them back up:
`docker-compose --env-file hosting.properties up -d`

Now, i'm a bit confused about this mix of docker-compose and docker setup.
Was under the impression that because we're using docker-compose based setup, docker shouldn't be needed.

Also, i noticed that after finally bringing up the containers, the earlier apps had been deleted. Whereas, they shouldn't have been because persistent volumes were used to create them.

Kindly advise on this.
Thanks.

**Comment by arladmin** [comment, marked as answer]
Found the command for the updating part:

`docker-compose --env-file hosting.properties pull && docker-compose --env-file hosting.properties up -d`

**Reply by shogunpurple** [reply]
@arladmin Sorry, just seeing this. The script above should update your environment.

When you run that command, are you losing all of your app data? We have a docker volume for couchDB which should contain (and persist) the data for your apps - I will investigate further and see if I can recreate this issue in a self hosted environment.

**Comment by Yakito** [comment]
Hi! Maybe someone over this thread can help. After a fresh install on DigitalOcean I was unable to set the tables to public (https://github.com/Budibase/budibase/discussions/3176). 
It was suggested to me to try to update Budi, but when I run
`budi hosting --update 
`
as suggested here https://docs.budibase.com/self-hosting/self-host I get

```
`Budibase Error
The Compose file './docker-compose.yaml' is invalid because:
services.watchtower-service.ports contains an invalid type, it should be a number, or an object
```
Any ideas what might be going on? 
Thanks a lot in advance!

**Reply by PGLange** [reply]
Hi Yakito,
I had the same problem and @shogunpurple explained that copiing your hosting hosting.properties to .env in the same directory will solve that issue.

[Issue#2501](https://github.com/Budibase/budibase/issues/2501)

**Reply by Yakito** [reply]
thanks a lot @PGLange , that worked!

**Reply by ovizii** [reply]
> Hi Yakito, I had the same problem and @shogunpurple explained that copiing your hosting hosting.properties to .env in the same directory will solve that issue.
> 
> [Issue#2501](https://github.com/Budibase/budibase/issues/2501)

I also had this problem. Followed the installation instructions and then got stuck. => https://docs.budibase.com/self-hosting/hosting-methods/docker-compose

With all of these variables not being able to be read. Your advice helped fix the problem but why does this happen and why is this not mentioned in the isntallation instructions?

```
docker-compose  logs -f                                                                                             
ERROR: The Compose file './docker-compose.yaml' is invalid because:                                                                                  
services.proxy-service.ports contains an invalid type, it should be a number, or an object                                                           
services.app-service.ports contains an invalid type, it should be a number, or an object                                                             
services.worker-service.ports contains an invalid type, it should be a number, or an object                                                          
services.couchdb-service.ports contains an invalid type, it should be a number, or an object                                                         
services.redis-service.ports contains an invalid type, it should be a number, or an object                                                           
services.watchtower-service.ports contains an invalid type, it should be a number, or an object                                                      
services.minio-service.ports contains an invalid type, it should be a number, or an object         
```

**Reply by shogunpurple** [reply]
@ovizii - the standard docker-compose installation instructions are correct. Due to the `hosting.properties`, you must pass the `--env-file` flag with every docker command. This is true with docker-compose in general when using an external env file.

`docker-compose --env-file hosting.properties logs -f`

I see what the issue is here. The standard docker-compose setup asks you to download the `hosting.properties` file, but the budi CLI uses `.env`. So if you start with the standard docker setup, then try to update with `budi hosting --update`, the `.env` file isn't present and the issue occurs.

What we could possibly do instead of `hosting.properties` is just ship a `.env` file by default for both types of installation. Docker picks this up automatically when you run `docker-compose up -d` as well, and it might save some confusion.

I'll update the docs to add an update command to the docker-compose docs.

**Reply by ovizii** [reply]
@shogunpurple sorry if I wasn't clear enough, I am just starting out, followed the docker-compose instructions which end up with: "docker-compose --env-file hosting.properties up" which leads to the errors I posted above. 

so, I used " --env-file hosting.properties" and still got these errors until I copied them to .env

(edit) I only replied to this thread about updating as it was the best occurrence of these errors I could find without opening my own thread.
